# Graded Challenge 3

Nama  : Rahardiansyah Fatoni

Batch : RMT-27

Graded Challenge 3 ini dibuat guna mengevaluasi konsep Web Scraping, Business Knowledge dan Practical Statistics sebagai berikut:

*   Mampu mengambil data dari sumber internet dengan Web Scraping
*   Mampu melakukan data preparation sebelum analisis selanjutnya
*   Mampu membangun problem statement dengan framework SMART sebagai salah satu langkah business understanding
*   Mampu melakukan perhitungan statistik deskriptif
*   Mampu melakukan pengujian statistik dan merumuskan hipotesis
*   Mampu mengambil insight/informasi dari hasil perhitungan
*   Mampu mengambil insight/informasi dari hasil perhitungan
*   Mampu mengambil kesimpulan yang menjawab problem statement

## A. Web Scraping

In [ ]:
# IMPORT MODULE UNTUK WEB SCRAPING
from bs4 import BeautifulSoup
from selenium.webdriver.firefox.options import Options
from selenium import webdriver

# PENAMBAHAN OPTIONS UNTUK WEBDRIVER UNTUK MEMPERCEPAT PROSES SCRAPE
options = Options()
options.add_argument('--disable-gpu')
options.add_argument('--headless')
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')

In [ ]:
# PEMBUATAN LIST UNTUK MENAMPUNG HASIL SCRAPE
list_nama_produk = []
list_harga_produk = []
list_penjual_produk = []
list_kota_produk = []
list_banyak_produk = []
list_rating_produk = []

In [ ]:
for page in range(1,11):
  driver = webdriver.Firefox(options=options)
  url=f"https://www.tokopedia.com/search?navsource=&page={page}&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st="
  driver.get(url)
  html = driver.page_source
  soup = BeautifulSoup(html, "html.parser")

  # soup.find_all KEPADA ELEMEN YANG MENCAKUP DATA: NAMA, PENJUAL, HARGA, BANYAK TERJUAL, KOTA PENJUAL, DAN RATING PRODUK
  for el in soup.find_all("a", {"class": "pcv3__info-content css-gwkf0u"}):
    # Use el.find to search within the context of the current loop iteration
    nama = el.find("div", {"class": "prd_link-product-name css-3um8ox"})
    list_nama_produk.append(nama.get_text() if nama else None)

    harga = el.find("div", {"class": "prd_link-product-price css-h66vau"})
    list_harga_produk.append(harga.get_text() if harga else None)

    penjual = el.find("span", {"class": "prd_link-shop-name css-1kdc32b flip"})
    list_penjual_produk.append(penjual.get_text() if penjual else None)

    kota = el.find("span", {"class": "prd_link-shop-loc css-1kdc32b flip"})
    list_kota_produk.append(kota.get_text() if kota else None)

    banyak = el.find("span", {"class": "prd_label-integrity css-1sgek4h"})
    list_banyak_produk.append(banyak.get_text() if banyak else None)

    rating = el.find("span", {"class": "prd_rating-average-text css-t70v7i"})
    list_rating_produk.append(rating.get_text() if rating else None)

## B. Data Preparation

In [ ]:
# IMPORT PANDAS
import pandas as pd

# PEMBUATAN DATAFRAME
df = pd.DataFrame()

# IMPORT DATA DARI LIST KE DALAM DATAFRAME
df['nama'] = list_nama_produk
df['harga'] = list_harga_produk
df['penjual'] = list_penjual_produk
df['kota'] = list_kota_produk
df['unit_terjual'] = list_banyak_produk
df['rating'] = list_rating_produk

In [ ]:
df

nama     harga  \
0    Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...  Rp13.000   
1                        Kylafood Seblak Original Play  Rp15.050   
2    Seblak Instan Ceu Nthien Khas Bandung Rasana N...  Rp17.000   
3                      Kylafood Seblak Rempah Autentik  Rp10.000   
4                   KERUPUK KRUPUK SEBLAK PEDAS 250 GR  Rp10.000   
..                                                 ...       ...   
345                      Seblak Instan Pedas Home Made   Rp3.500   
346  Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...  Rp13.000   
347  seblak kering pedas 250 gr/seblak kencur/kerup...   Rp9.999   
348  keripik beling/kaca 1/2 kg. bumbu seblak daun ...  Rp25.999   
349  Gelifood Combo 3pcs Seblak Instan Kerupuk Mawa...  Rp35.000   

                      penjual               kota  unit_terjual rating  
0                    Lidigeli         Kab. Garut  250+ terjual    4.8  
1                    kylafood            Bandung  1rb+ terjual    4.9  
2    Central Seblak Nusantara  Tangerang Selatan  2rb+ terjual    4.9  
3            Kylafood Jakarta    Jakarta Selatan  100+ terjual    5.0  
4             TOKO KUE MEISYA      Jakarta Pusat  100+ terjual    4.9  
..                        ...                ...           ...    ...  
345                the Dhecip  Tangerang Selatan  2rb+ terjual    4.8  
346                  Lidigeli         Kab. Garut  250+ terjual    4.8  
347                   aghniya  Tangerang Selatan  500+ terjual    4.8  
348                   aghniya  Tangerang Selatan  250+ terjual    4.9  
349                  Lidigeli         Kab. Garut  100+ terjual    4.9  

[350 rows x 6 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   nama          350 non-null    object
 1   harga         350 non-null    object
 2   penjual       350 non-null    object
 3   kota          350 non-null    object
 4   unit_terjual  343 non-null    object
 5   rating        339 non-null    object
dtypes: object(6)
memory usage: 16.5+ KB


INSIGHT: terdapat 350 data non-null. Langkah selanjutnya adalah membersihkan data.

In [ ]:
# MENGGUNAKAN REPLACE UNTUK MENGHILANGKAN "Rp" dan "." DARI KOLOM <harga>
df['harga'] = df['harga'].str.replace("Rp", "").str.replace(".","")

# MENGGUNAKAN REPLACE UNTUK MENGHILANGKAN "+ terjual" , "rb", dan " terjual" DARI KOLOM unit_terjual
df['unit_terjual'] = df['unit_terjual'].str.replace("\+ terjual","").str.replace("rb","000").str.replace(" terjual","")

# MENGUBAH DATA TYPE MENJADI NUMERIKAL
df['harga'] = pd.to_numeric(df['harga'])
df['unit_terjual'] = pd.to_numeric(df['unit_terjual'])
df['rating'] = pd.to_numeric(df['rating'])

Sekarang tipe data yang ada adalah:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   nama          350 non-null    object 
 1   harga         350 non-null    int64  
 2   penjual       350 non-null    object 
 3   kota          350 non-null    object 
 4   unit_terjual  343 non-null    float64
 5   rating        339 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 16.5+ KB


Karena sudah sesuai, dapat dilanjutkan ke tahap selanjutnya!

In [ ]:
df.to_csv("seblak_tokopedia.csv", index = False)

In [6]:
import pandas as pd

df = pd.read_csv("/content/seblak_tokopedia.csv")

## C. Business Understanding/Problem Statement

Problem Statement:

Tujuan dari pengamatan ini adalah untuk mengetahui dan menganalisis korelasi antara harga produk seblak di Tokopedia dengan lokasi penjualan, dan rating menggunakan SMART framework:

**Specific**: Mengetahui dan menganalisis korelasi harga produk seblak di Tokopedia terhadap lokasi penjualan dan rating.

**Measurable**: Mendapat insight tentang seberapa berpengaruhnya variabel harga
 terhadap rating terhadap produk seblak.

**Achievable**: Menggunakan data yang didapatkan melalui scraping dari Tokopedia.

**Relevant**: Mengetahui range harga yang meningkatkan kemungkinan pembelian akan berguna kedepannya untuk dijadikan referensi dalam memberikan harga produk.

**Time-Bound**: Analisis akan dilakukan dalam 3 hari.

## D. Analysis

### Untuk mengawali analisis kita akan mencari tau beberapa nilai krusial seperti rata-rata, median, dan modus dari kolom yang relevan.

1. Data harga:

In [14]:
print(f"Rata-rata harga : {df['harga'].mean()}")
print(f"Median harga : {df['harga'].median()}")
print(f"Modus harga : {df['harga'].mode()[0]}")
print(f"Standar deviasi harga : {df['harga'].std()}")

Rata-rata harga : 16966.47142857143
Median harga : 13000.0
Modus harga : 25000
Standar deviasi harga : 15907.213765458222


In [9]:
print(f"Skewness harga: {df['harga'].skew()}")

Skewness harga: 3.997388803443054


Karena data harga memiliki skewness diatas 1, maka dapat disimpulkan data memiliki distribusi positif skew atau mencondong ke kanan.

In [10]:
print(f"Kurtosis harga: {df['harga'].kurtosis()}")

Kurtosis harga: 29.455604730580966


Karena data harga memiliki kurtosis diatas 0, maka tipe kurtosis nya adalah Leptokurtic.

2. Data rating:

In [16]:
print(f"Rata-rata rating : {df['rating'].mean()}")
print(f"Median rating : {df['rating'].median()}")
print(f"Modus rating : {df['rating'].mode()[0]}")
print(f"Standar deviasi rating  : {df['rating'].std()}")

Rata-rata rating : 4.8855457227138634
Median rating : 4.9
Modus rating : 4.9
Standar deviasi rating  : 0.09698371182249955


In [11]:
print(f"Skewness rating: {df['rating'].skew()}")

Skewness rating: -1.7034350710344979


Karena data rating memiliki skewness diatas -1, maka dapat disimpulkan data memiliki distribusi positif skew atau mencondong ke kiri.

In [12]:
print(f"Kurtosis rating: {df['rating'].kurtosis()}")

Kurtosis rating: 9.333766351013523


Karena data rating memiliki kurtosis diatas 0, maka tipe kurtosis nya adalah Leptokurtic.

3. Data unit terjual:

In [18]:
print(f"Rata-rata unit terjual : {df['unit_terjual'].mean()}")
print(f"Median unit terjual : {df['unit_terjual'].median()}")
print(f"Modus unit terjual : {df['unit_terjual'].mode()[0]}")
print(f"Standar deviasi unit terjual  : {df['unit_terjual'].std()}")

Rata-rata unit terjual : 1042.5597667638483
Median unit terjual : 250.0
Modus unit terjual : 250.0
Standar deviasi unit terjual  : 2417.286900762666


In [19]:
print(f"Skewness unit terjual: {df['unit_terjual'].skew()}")

Skewness unit terjual: 3.1684804420092947


Karena data unit terjual memiliki skewness diatas 1, maka dapat disimpulkan data memiliki distribusi positif skew atau mencondong ke kanan.

In [20]:
print(f"Kurtosis unit terjual: {df['unit_terjual'].kurtosis()}")

Kurtosis unit terjual: 8.890265745789009


Karena data unit terjual memiliki kurtosis diatas 0, maka tipe kurtosis nya adalah Leptokurtic.

### Sekarang, kita akan mencari tahu berapa potensi maksimal dan minimal dari penjualan seblak per bulan di Tokopedia.

In [ ]:
q1 = df['harga'].quantile(0.25)
q3 = df['harga'].quantile(0.75)
iqr = q3 - q1

# Batasan menggunakan Tukey's Rule
up_bound = q3 + 1.5 * iqr
low_bound = q1 - 1.5 * iqr

print(f'Upper Boundary: {up_bound:.2f}')
print(f'Lower Boundary: {low_bound:.2f}')

# Pengecekan outlier
outliers = df[(df['harga'] > up_bound) | (df['harga'] < low_bound)]
print(f'Jumlah outliers: {len(outliers)}')

Upper Boundary: 51250.00
Lower Boundary: -18750.00
Jumlah outliers: 7


Asumsi: Data terdistribusi normal, informasi produk terjual merupakan penjualan produk per bulan, dan confidence level 95%.

In [ ]:
df['pendapatan'] = df['harga']*df['unit_terjual']

In [ ]:
# Mencari rata-rata dan standar deviasi pendapatan
mean_pendapatan = df['pendapatan'].mean()
std_dev_pendapatan = df['pendapatan'].std()
print(f'mean pendapatan: {mean_pendapatan}')
print(f'std deviasi pendapatan: {std_dev_pendapatan}')

mean pendapatan: 6760155.099125365
std deviasi pendapatan: 11837343.595173921


In [23]:
# IMPORT MODULE NUMPY DAN STATS DARI SCIPY
import numpy as np
from scipy import stats

`stats.norm.interval(conf_level,loc=average,scale="The_width_of th_ distribution_from_the_average")`

The width of the distribution from the average = `std/sqrt(N)`

In [ ]:
std = std_dev_pendapatan
N = len(df['pendapatan'])
low, up = stats.norm.interval(0.95,loc=df['pendapatan'].mean(),scale=std/np.sqrt(N))
print('Lower Limit:',low)
print('Upper Limit:',up)

Lower Limit: 5520021.932407474
Upper Limit: 8000288.265843255


###**Insight: Maka potensi pendapatan minimun penjualan seblak dalam sebulan Rp 5,5 - 8 jt.**

### Sekarang, memacu pada SMART framework diawal, saatnya mencari tahu apakah ada korelasi antara harga dengan lokasi penjualan.


In [25]:
# Membuat dataframe khusus untuk daerah Jabodetabek.
daerah_jabodetabek = ['Jakarta', 'Tangerang', 'Bogor', 'Depok', 'Bekasi']
harga_jabodetabek = df[df['kota'].isin(daerah_jabodetabek)]['harga']
# ~ Dapat digunakan untuk mencari harga diluar Jabodetabek.
harga_luar = df[~df['kota'].isin(daerah_jabodetabek)]['harga']

df_jabodetabek = pd.DataFrame(harga_jabodetabek)
df_luar = pd.DataFrame(harga_luar)

*Two sample test* akan digunakan untuk membandingkan harga di Jabodetabek dengan harga diluar Jabodetabek, dikaremakan data yang akan dibandingkan adalah dua sampel.

In [ ]:
print(f'Rata-rata harga seblak di Jabodetabek: {df_jabodetabek.mean()["harga"]:.2f}')
print(f'Rata-rata harga seblak di luar Jabodetabek: {df_luar.mean()["harga"]:.2f}')

Rata-rata harga seblak di Jabodetabek: 19696.50
Rata-rata harga seblak di luar Jabodetabek: 16424.21


$H_0$: Rata-rata pendapatan penjualan seblak di Jabodetabek == rata-rata pendapatan penjualan seblak di luar Jabodetabek

$H_1$: Rata-rata pendapatan penjualan seblak di Jabodetabek != rata-rata pendapatan penjualan seblak di luar Jabodetabek

In [26]:
t_stat, p_val = stats.ttest_ind(df_jabodetabek['harga'],df_luar['harga'])
print('T-Statistic:',t_stat)
print('P-value:',p_val)

T-Statistic: 1.4331260296200368
P-value: 0.15271958003691344


Dikarenakan P-value > 0.05, maka H0 tidak ditolak, atau dapat dibilang bahwa tidak ada perbedaan signifikan dalam harga barang antara Jabodetabek dan di luar Jabodetabek.

### Selanjutnya kita akan mencari tahu apakah ada korelasi dari harga terhadap rating produk.
Untuk itu kita akan menggunakan Chi-Squared test.

$H_0$: Tidak hubungan harga dengan rating produk.

$H_1$: Ada hubungan harga dengan rating produk.

In [ ]:
#Contingency Table
contingency_table = pd.crosstab(df['harga'],df['rating'])

In [ ]:
res = stats.chi2_contingency(contingency_table)
print("P-value:",res.pvalue)
if res.pvalue < 0.05:
  print("Dependen.")

P-value: 1.264833028201808e-139
Dependen.


P-value < 0.05 oleh karena itu H0 ditulak, yang berarti ada hubungan antara harga dengan rating produk (dependen)

## E. Conclusion

1. Rata-rata harga seblak adalah Rp 16.966.
2. Rata-rata pendapatan potensial tiap bulan penjualan seblak online adalah Rp 6.760.155, dengan minimum Rp 5.520.021 dan maksimum Rp 8.000.288.
3. Tidak ada perbedaan signifikan terhadap harga seblak di Jabodetabek dan diluar Jabodetabek.
4. Ada hubungan antara harga dengan rating produk.